File: src/receipt_canonicalization.py
"""Versioned canonical receipt serialization profiles."""

from __future__ import annotations

import hashlib
import json
import math
import re
import unicodedata
from collections.abc import Mapping
from datetime import datetime, timezone
from decimal import Decimal, InvalidOperation, ROUND_HALF_UP
from typing import Any

CANONICALIZATION_PROFILE_V1 = "poc-c14n-v1"
HASH_ALGORITHM = "sha-256"

TOP_LEVEL_FIELDS = frozenset({
    "version", "schema_version", "canonicalization_profile", "hash_algorithm",
    "session_type", "session_id", "user_id", "evse_id", "ocpp_tx_id",
    "start_ts", "end_ts", "energy_kwh", "energy_summary", "pricing",
    "settlement", "merkle_root", "stream_hash_alg",
})
ENERGY_SUMMARY_FIELDS = frozenset({"import_kwh", "export_kwh", "net_kwh"})
PRICING_FIELDS = frozenset({"currency", "model", "components", "import_components", "export_components"})
PRICING_COMPONENT_FIELDS = frozenset({"from", "to", "price_per_kwh"})
SETTLEMENT_FIELDS = frozenset({"gross_import_cost", "gross_export_credit", "net_amount", "currency"})

ENERGY_PATHS = {
    ("energy_kwh",),
    ("energy_summary", "import_kwh"),
    ("energy_summary", "export_kwh"),
    ("energy_summary", "net_kwh"),
}
MONEY_PATHS = {
    ("settlement", "gross_import_cost"),
    ("settlement", "gross_export_credit"),
    ("settlement", "net_amount"),
}
TIMESTAMP_NAMES = {"start_ts", "end_ts", "from", "to"}


class CanonicalizationError(ValueError):
    pass


def canonicalize_receipt(receipt: Mapping[str, Any], profile: str) -> bytes:
    """Return the exact UTF-8 hash input for a supported profile."""
    if profile != CANONICALIZATION_PROFILE_V1:
        raise CanonicalizationError(f"Unknown canonicalization profile: {profile}")
    if not isinstance(receipt, Mapping):
        raise CanonicalizationError("Receipt must be an object")
    _validate_profile_shape(receipt)
    if receipt.get("canonicalization_profile") != profile:
        raise CanonicalizationError("Receipt canonicalization_profile does not match requested profile")
    if receipt.get("hash_algorithm") != HASH_ALGORITHM:
        raise CanonicalizationError(f"hash_algorithm must be {HASH_ALGORITHM}")
    normalized = _normalize(dict(receipt), ())
    text = json.dumps(
        normalized,
        ensure_ascii=False,
        allow_nan=False,
        sort_keys=True,
        separators=(",", ":"),
    )
    try:
        return text.encode("utf-8")
    except UnicodeEncodeError as exc:
        raise CanonicalizationError("Strings must not contain lone Unicode surrogates") from exc


def hash_canonical_receipt(receipt: Mapping[str, Any], profile: str) -> str:
    return "0x" + hashlib.sha256(canonicalize_receipt(receipt, profile)).hexdigest()


def parse_json_strict(payload: str | bytes) -> Any:
    """Parse JSON while rejecting duplicate object keys and non-finite constants."""
    if isinstance(payload, bytes):
        if payload.startswith(b"\xef\xbb\xbf"):
            raise CanonicalizationError("UTF-8 BOM is not allowed")
        payload = payload.decode("utf-8")

    def object_pairs(pairs):
        result = {}
        for key, value in pairs:
            if key in result:
                raise CanonicalizationError(f"Duplicate JSON object key: {key}")
            result[key] = value
        return result

    def reject_constant(value):
        raise CanonicalizationError(f"Non-finite JSON number is not allowed: {value}")

    return json.loads(payload, object_pairs_hook=object_pairs, parse_constant=reject_constant)


def _validate_profile_shape(receipt: Mapping[str, Any]) -> None:
    _exact_fields(receipt, TOP_LEVEL_FIELDS, "receipt")
    _object_with_fields(receipt["energy_summary"], ENERGY_SUMMARY_FIELDS, "energy_summary")
    pricing = _object_with_fields(receipt["pricing"], PRICING_FIELDS, "pricing")
    for list_name in ("components", "import_components", "export_components"):
        components = pricing[list_name]
        if not isinstance(components, list):
            raise CanonicalizationError(f"pricing.{list_name} must be an array")
        for index, component in enumerate(components):
            _object_with_fields(component, PRICING_COMPONENT_FIELDS, f"pricing.{list_name}[{index}]")
    _object_with_fields(receipt["settlement"], SETTLEMENT_FIELDS, "settlement")


def _object_with_fields(value: Any, fields: frozenset[str], path: str) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise CanonicalizationError(f"{path} must be an object")
    _exact_fields(value, fields, path)
    return value


def _exact_fields(value: Mapping[str, Any], fields: frozenset[str], path: str) -> None:
    actual = set(value)
    missing = sorted(fields - actual)
    extra = sorted(actual - fields)
    if missing:
        raise CanonicalizationError(f"{path} missing required fields: {', '.join(missing)}")
    if extra:
        raise CanonicalizationError(f"{path} contains fields not defined by {CANONICALIZATION_PROFILE_V1}: {', '.join(extra)}")
    nulls = sorted(key for key, item in value.items() if item is None)
    if nulls:
        raise CanonicalizationError(f"{path} fields are not nullable: {', '.join(nulls)}")


def _normalize(value: Any, path: tuple[str, ...]) -> Any:
    if isinstance(value, Mapping):
        result = {}
        for key, item in value.items():
            if not isinstance(key, str):
                raise CanonicalizationError(f"Object key must be a string at {_path(path)}")
            normalized_key = _normalize_string(key)
            if normalized_key in result:
                raise CanonicalizationError(f"Unicode-normalized duplicate key: {normalized_key}")
            result[normalized_key] = _normalize(item, path + (key,))
        return result
    if isinstance(value, list):
        return [_normalize(item, path + ("[]",)) for item in value]
    if value is None:
        raise CanonicalizationError(f"Null is not allowed at {_path(path)}")
    if path in ENERGY_PATHS or path in MONEY_PATHS:
        return _fixed_decimal(value, 3, path)
    if len(path) >= 3 and path[-1] == "price_per_kwh" and path[-2] == "[]":
        return _fixed_decimal(value, 3, path)
    if path and path[-1] in TIMESTAMP_NAMES:
        return _timestamp(value, path)
    if isinstance(value, str):
        return _normalize_string(value)
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float, Decimal)):
        if isinstance(value, float) and not math.isfinite(value):
            raise CanonicalizationError(f"Non-finite number at {_path(path)}")
        raise CanonicalizationError(f"Unscaled numeric field is not allowed at {_path(path)}")
    raise CanonicalizationError(f"Unsupported value type at {_path(path)}: {type(value).__name__}")


def _fixed_decimal(value: Any, scale: int, path: tuple[str, ...]) -> str:
    if not isinstance(value, str):
        raise CanonicalizationError(f"Canonical decimal must be a string at {_path(path)}")
    if not re.fullmatch(r"[+-]?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?", value):
        raise CanonicalizationError(f"Invalid decimal at {_path(path)}: {value}")
    try:
        decimal_value = Decimal(str(value))
    except (InvalidOperation, ValueError) as exc:
        raise CanonicalizationError(f"Invalid decimal at {_path(path)}: {value}") from exc
    if not decimal_value.is_finite():
        raise CanonicalizationError(f"Non-finite number at {_path(path)}")
    quantum = Decimal(1).scaleb(-scale)
    rounded = decimal_value.quantize(quantum, rounding=ROUND_HALF_UP)
    if rounded == 0:
        rounded = abs(rounded)
    return format(rounded, f".{scale}f")


def _timestamp(value: Any, path: tuple[str, ...]) -> str:
    if not isinstance(value, str):
        raise CanonicalizationError(f"Timestamp must be a string at {_path(path)}")
    if not re.fullmatch(
        r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{1,6})?(?:Z|z|[+-]\d{2}:\d{2})",
        value,
    ):
        raise CanonicalizationError(f"Invalid RFC 3339 timestamp at {_path(path)}: {value}")
    raw = value[:-1] + "+00:00" if value.endswith(("Z", "z")) else value
    try:
        parsed = datetime.fromisoformat(raw)
    except ValueError as exc:
        raise CanonicalizationError(f"Invalid RFC 3339 timestamp at {_path(path)}: {value}") from exc
    if parsed.tzinfo is None or parsed.utcoffset() is None:
        raise CanonicalizationError(f"Naive timestamp is not allowed at {_path(path)}")
    utc = parsed.astimezone(timezone.utc)
    return utc.strftime("%Y-%m-%dT%H:%M:%S.%fZ")


def _normalize_string(value: str) -> str:
    normalized = unicodedata.normalize("NFC", value)
    if any(0xD800 <= ord(char) <= 0xDFFF for char in normalized):
        raise CanonicalizationError("Strings must not contain lone Unicode surrogates")
    return normalized


def _path(path: tuple[str, ...]) -> str:
    return ".".join(path).replace(".[]", "[]") or "receipt"

File: scripts/receipt_canonicalization.mjs
#!/usr/bin/env node
/** Independent JavaScript reference for poc-c14n-v1. */

import crypto from "node:crypto";
import fs from "node:fs";

export const PROFILE = "poc-c14n-v1";
export const HASH_ALGORITHM = "sha-256";

const TOP = new Set(["version","schema_version","canonicalization_profile","hash_algorithm","session_type","session_id","user_id","evse_id","ocpp_tx_id","start_ts","end_ts","energy_kwh","energy_summary","pricing","settlement","merkle_root","stream_hash_alg"]);
const ENERGY_SUMMARY = new Set(["import_kwh","export_kwh","net_kwh"]);
const PRICING = new Set(["currency","model","components","import_components","export_components"]);
const COMPONENT = new Set(["from","to","price_per_kwh"]);
const SETTLEMENT = new Set(["gross_import_cost","gross_export_credit","net_amount","currency"]);
const ENERGY_PATHS = new Set(["energy_kwh","energy_summary.import_kwh","energy_summary.export_kwh","energy_summary.net_kwh"]);
const MONEY_PATHS = new Set(["settlement.gross_import_cost","settlement.gross_export_credit","settlement.net_amount"]);

export function canonicalizeReceipt(receipt, profile) {
  if (profile !== PROFILE) throw new Error(`Unknown canonicalization profile: ${profile}`);
  if (!isObject(receipt)) throw new Error("Receipt must be an object");
  validateShape(receipt);
  if (receipt.canonicalization_profile !== profile) throw new Error("Receipt canonicalization_profile does not match requested profile");
  if (receipt.hash_algorithm !== HASH_ALGORITHM) throw new Error(`hash_algorithm must be ${HASH_ALGORITHM}`);
  const text = JSON.stringify(normalize(receipt, []));
  return Buffer.from(text, "utf8");
}

export function hashCanonicalReceipt(receipt, profile) {
  return `0x${crypto.createHash("sha256").update(canonicalizeReceipt(receipt, profile)).digest("hex")}`;
}

function validateShape(receipt) {
  exact(receipt, TOP, "receipt");
  exactObject(receipt.energy_summary, ENERGY_SUMMARY, "energy_summary");
  exactObject(receipt.pricing, PRICING, "pricing");
  for (const name of ["components", "import_components", "export_components"]) {
    const values = receipt.pricing[name];
    if (!Array.isArray(values)) throw new Error(`pricing.${name} must be an array`);
    values.forEach((item, index) => exactObject(item, COMPONENT, `pricing.${name}[${index}]`));
  }
  exactObject(receipt.settlement, SETTLEMENT, "settlement");
}

function exactObject(value, fields, path) {
  if (!isObject(value)) throw new Error(`${path} must be an object`);
  exact(value, fields, path);
}

function exact(value, fields, path) {
  const keys = Object.keys(value);
  const missing = [...fields].filter(key => !Object.hasOwn(value, key)).sort(codePointCompare);
  const extra = keys.filter(key => !fields.has(key)).sort(codePointCompare);
  if (missing.length) throw new Error(`${path} missing required fields: ${missing.join(", ")}`);
  if (extra.length) throw new Error(`${path} contains fields not defined by ${PROFILE}: ${extra.join(", ")}`);
  const nulls = keys.filter(key => value[key] === null).sort(codePointCompare);
  if (nulls.length) throw new Error(`${path} fields are not nullable: ${nulls.join(", ")}`);
}

function normalize(value, path) {
  if (Array.isArray(value)) return value.map(item => normalize(item, [...path, "[]"]));
  if (isObject(value)) {
    const result = {};
    const normalizedKeys = new Set();
    for (const key of Object.keys(value).sort(codePointCompare)) {
      const normalizedKey = normalizeString(key);
      if (normalizedKeys.has(normalizedKey)) throw new Error(`Unicode-normalized duplicate key: ${normalizedKey}`);
      normalizedKeys.add(normalizedKey);
      result[normalizedKey] = normalize(value[key], [...path, key]);
    }
    return result;
  }
  if (value === null) throw new Error(`Null is not allowed at ${showPath(path)}`);
  const joined = path.join(".");
  if (ENERGY_PATHS.has(joined) || MONEY_PATHS.has(joined) || (path.at(-1) === "price_per_kwh" && path.at(-2) === "[]")) return fixedDecimal(value, 3, path);
  if (["start_ts", "end_ts", "from", "to"].includes(path.at(-1))) return timestamp(value, path);
  if (typeof value === "string") return normalizeString(value);
  if (typeof value === "boolean") return value;
  if (typeof value === "number" || typeof value === "bigint") {
    if (typeof value === "number" && !Number.isFinite(value)) throw new Error(`Non-finite number at ${showPath(path)}`);
    throw new Error(`Unscaled numeric field is not allowed at ${showPath(path)}`);
  }
  throw new Error(`Unsupported value type at ${showPath(path)}: ${typeof value}`);
}

function fixedDecimal(value, scale, path) {
  if (typeof value !== "string") throw new Error(`Canonical decimal must be a string at ${showPath(path)}`);
  let raw = value;
  const match = raw.match(/^([+-]?)(\d+)(?:\.(\d*))?(?:[eE]([+-]?\d+))?$/);
  if (!match) throw new Error(`Invalid decimal at ${showPath(path)}: ${value}`);
  const negative = match[1] === "-";
  const integer = match[2];
  const fraction = match[3] || "";
  const exponent = Number(match[4] || 0);
  let digits = (integer + fraction).replace(/^0+(?=\d)/, "");
  let decimalPlaces = fraction.length - exponent;
  if (decimalPlaces < 0) { digits += "0".repeat(-decimalPlaces); decimalPlaces = 0; }
  let scaled;
  if (decimalPlaces <= scale) {
    scaled = BigInt(digits || "0") * 10n ** BigInt(scale - decimalPlaces);
  } else {
    const divisor = 10n ** BigInt(decimalPlaces - scale);
    const absolute = BigInt(digits || "0");
    scaled = absolute / divisor;
    const remainder = absolute % divisor;
    if (remainder * 2n >= divisor) scaled += 1n; // ROUND_HALF_UP
  }
  if (negative && scaled !== 0n) scaled = -scaled;
  const sign = scaled < 0n ? "-" : "";
  let output = (scaled < 0n ? -scaled : scaled).toString().padStart(scale + 1, "0");
  return `${sign}${output.slice(0, -scale)}.${output.slice(-scale)}`;
}

function timestamp(value, path) {
  if (typeof value !== "string") throw new Error(`Timestamp must be a string at ${showPath(path)}`);
  const match = value.match(/^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{1,6}))?(Z|z|[+-]\d{2}:\d{2})$/);
  if (!match) throw new Error(`Invalid RFC 3339 timestamp at ${showPath(path)}: ${value}`);
  const [, year, month, day, hour, minute, second, fraction = "", zone] = match;
  const micros = Number(fraction.padEnd(6, "0"));
  let offsetMinutes = 0;
  if (!["Z", "z"].includes(zone)) {
    const sign = zone[0] === "-" ? -1 : 1;
    offsetMinutes = sign * (Number(zone.slice(1, 3)) * 60 + Number(zone.slice(4, 6)));
  }
  const localMillis = Date.UTC(Number(year), Number(month) - 1, Number(day), Number(hour), Number(minute), Number(second));
  const local = new Date(localMillis);
  if (local.getUTCFullYear() !== Number(year) || local.getUTCMonth() !== Number(month) - 1 || local.getUTCDate() !== Number(day) || local.getUTCHours() !== Number(hour) || local.getUTCMinutes() !== Number(minute) || local.getUTCSeconds() !== Number(second)) throw new Error(`Invalid RFC 3339 timestamp at ${showPath(path)}: ${value}`);
  const millis = localMillis - offsetMinutes * 60000;
  const date = new Date(millis);
  if (Number.isNaN(date.valueOf())) throw new Error(`Invalid RFC 3339 timestamp at ${showPath(path)}: ${value}`);
  const base = date.toISOString().slice(0, 19);
  return `${base}.${String(micros).padStart(6, "0")}Z`;
}

function normalizeString(value) {
  for (const char of value) {
    const code = char.codePointAt(0);
    if (code >= 0xD800 && code <= 0xDFFF) throw new Error("Strings must not contain lone Unicode surrogates");
  }
  return value.normalize("NFC");
}

function codePointCompare(left, right) {
  const a = Array.from(left, char => char.codePointAt(0));
  const b = Array.from(right, char => char.codePointAt(0));
  for (let i = 0; i < Math.min(a.length, b.length); i++) if (a[i] !== b[i]) return a[i] - b[i];
  return a.length - b.length;
}

function showPath(path) { return path.join(".").replaceAll(".[]", "[]") || "receipt"; }
function isObject(value) { return value !== null && typeof value === "object" && !Array.isArray(value); }

if (import.meta.url === `file://${process.argv[1]}`) {
  const input = JSON.parse(fs.readFileSync(process.argv[2] || 0, "utf8"));
  const bytes = canonicalizeReceipt(input, input.canonicalization_profile);
  process.stdout.write(JSON.stringify({canonical_utf8_base64: bytes.toString("base64"), canonical_text: bytes.toString("utf8"), hash: hashCanonicalReceipt(input, input.canonicalization_profile)}));
}
